internal/pkg/manifest/transform.go

```diff
@@ -19,89 +42,281 @@ var basicKinds = []reflect.Kind{
-func (t imageTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
-	if typ == reflect.TypeOf(BuildArgsOrString{}) {
-		return transformBuildArgsOrString()
+// Transformer returns custom merge logic for HealthCheckArgsOrString's fields.
+func (t healthCheckArgsOrStringTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ != reflect.TypeOf(HealthCheckArgsOrString{}) {
+		return nil
+	}
+
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(HealthCheckArgsOrString), src.Interface().(HealthCheckArgsOrString)
+
+		if srcStruct.HealthCheckPath != nil {
+			dstStruct.HealthCheckArgs = HTTPHealthCheckArgs{}
+		}
+
+		if !srcStruct.HealthCheckArgs.isEmpty() {
+			dstStruct.HealthCheckPath = nil
+		}
+
+		if dst.CanSet() { // For extra safety to prevent panicking.
+			dst.Set(reflect.ValueOf(dstStruct))
+		}
+		return nil
 	}
-	return transformBasic(typ)
 }
 
-// transformBasic implements customized merge logic for manifest fields that are number, string, bool, array, and duration.
-func transformBasic(typ reflect.Type) func(dst, src reflect.Value) error {
-	if typ.Kind() == reflect.Slice {
-		return transformSlice()
+type countTransformer struct{}
+
+// Transformer returns custom merge logic for Count's fields.
+func (t countTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ != reflect.TypeOf(Count{}) {
+		return nil
 	}
 
-	if typ.Kind() == reflect.Ptr {
-		for _, k := range basicKinds {
-			if typ.Elem().Kind() == k {
-				return transformPBasic()
-			}
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(Count), src.Interface().(Count)
+
+		if !srcStruct.AdvancedCount.IsEmpty() {
+			dstStruct.Value = nil
+		}
+
+		if srcStruct.Value != nil {
+			dstStruct.AdvancedCount = AdvancedCount{}
 		}
 
-		if typ.Elem().Kind() == reflect.Struct {
-			return transformPStruct()
+		if dst.CanSet() { // For extra safety to prevent panicking.
+			dst.Set(reflect.ValueOf(dstStruct))
 		}
+		return nil
+	}
+}
+
+type advancedCountTransformer struct{}
+
+// Transformer returns custom merge logic for AdvancedCount's fields.
+func (t advancedCountTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	if typ != reflect.TypeOf(AdvancedCount{}) {
+		return nil
+	}
+
+	return func(dst, src reflect.Value) error {
+		dstStruct, srcStruct := dst.Interface().(AdvancedCount), src.Interface().(AdvancedCount)
+
+		spotSpecified := srcStruct.Spot != nil
+		elseSpecified := srcStruct.Range != nil ||
+			srcStruct.CPU != nil ||
+			srcStruct.Memory != nil ||
+			srcStruct.Requests != nil ||
+			srcStruct.ResponseTime != nil
+		if spotSpecified && elseSpecified {
+			return fmt.Errorf(fmtExclusiveFieldsSpecifiedTogether, "spot", "is",
+				english.WordSeries([]string{"range", "cpu", "memory", "requests", "response_time"}, "and"))
+		}
```

Comment:
can we not use: `srcStruct.IsValid()` I believe this behavior is already implemented.

I think in general:
1. In `transformers` we should assume that both `dstStruct` and `srcStruct` are unmarshaled properly. This is the responsibility of `UnmarshalYAML`: https://github.com/aws/copilot-cli/blob/fe9cefdc82f3cc3a2d7cc5ceb60a414e1db3b9e2/internal/pkg/manifest/svc.go#L126


2. The transformer should only be used to unset fields like the if-statements below.

Reply:
Agreed. I also thought of this as a temporary resolution before, say a `manifest.validation` package is implemented. 

Will move the logic to `UnmarshalYAML`!